tiatoolbox/models/dataset/interactive_segmentation.py

```diff
@@ -0,0 +1,156 @@
+
+        Examples:
+            >>> # create a dataset to extract small patches around each point on a patch image
+            >>> ds = InteractiveSegmentorDataset(
+            ...     img_path = 'example_image.png',
+            ...     points = 'example_points.csv',
+            ...     mode = 'patch',         
+            ... )
+
+        """
+        # Not using IOConfig at the moment
+
+        super().__init__()
+
+        if not os.path.isfile(img_path):
+            raise ValueError("`img_path` must be a valid file path.")
+        if mode not in ["patch", "wsi", "tile"]:
+            raise ValueError(f"`{mode}` is not supported.")
+
+        self.img_path = img_path
+        self.label = label
+        self.mode = mode
+        self.patch_size = patch_size    
+        self.resolution = resolution    
+        self.units = units             
+
+        # Read the points('clicks') into a panda df
+        self.locations = read_locations(points)
+
+        self.patch_extractor = get_patch_extractor("point",  
+            input_img = self.img_path, locations_list = points, patch_size=self.patch_size,
+            resolution = self.resolution, units = self.units)
+
+    def __getitem__(self, idx):
+        patch = self.patch_extractor.__getitem__(idx)
+
+        boundingBox = self.get_boundingBox(idx)
+
+        # we know nucPoint is at the centre of the patch:
+        nucPoint = np.zeros((1, self.patch_size[1], self.patch_size[1]), dtype=np.uint8)
+        nucPoint[0,int((self.patch_size[1]-1)/2),int((self.patch_size[1]-1)/2)] = 1
+
+        exclusionMap = self.get_exclusionMap(idx, boundingBox)
+
+        patch = np.moveaxis(patch, 2, 0)
+        patch = patch / 255
+
+        input = np.concatenate((patch, nucPoint, exclusionMap), axis=0, dtype=np.float32)   # shape=(c=5,h,w)
+ 
+        data = {
+            "input": input,
+            "boundingBox": boundingBox,
+            "click": (self.locations["x"][idx], self.locations["y"][idx])
+        }
+        if self.label is not None:
+            data["label"] = self.label
+        
+        return data
+
+
+    def get_boundingBox(self, idx):
+        """This function returns a bounding box of size (patch_size x patch_size) that has the click as its centre.
+            The bounding box is the same box that is used in patch extraction.
+
+        Args:
+            idx (int): The index of the point ("Click") to get a bounding box for.
+        Returns:
+            bounds: a list of coordinates in `[start_x, start_y, end_x, end_y]`
+            format to be used for patch extraction.
+
+        """
+
+        #Coordinates of the top left corner of each patch:
+        location = (self.patch_extractor.locations_df["x"][idx], self.patch_extractor.locations_df["y"][idx])
+
+        tl = np.array(location)
+        br = location + np.array(self.patch_size)
+        bounds = np.concatenate([tl, br])
+
+        return  bounds
+    
+    def get_exclusionMap(self, idx, boundingBox):
```

Comment:
This function has a good property that it does not need saving a big array in RAM but its computationally extensive. Imagine we have 100000 points in the point list, each time that `__get_item__` method is called, this function also calls a for loop over all 99999 other cells, which is not optimized I believe. There are some ways to avoid this:

1- Use the original NuClick approach: creating a big array of `click_point` and crop the bounding box region from it with each __get_item__ . This way you get rid of the internal for loop and it's fast. Disadvantage is that it might require at max 500 MB of RAM for big whole slide image. (which I'm fine with)

2- This approach is to try vectorizing the inner for loop. Instead of going over all `otherPoints` in a for loop, try using array filtering, for example:
```
xy_locations = np.concat([x_locations, y_locations], axis=1)
sel = np.any(xy_locations  > boundingBox[:2], axis=1)
sel |= np.any(xy_locations  < boundingBox[2:], axis=1)
filtered_xy_locs = xy_locations[sel]

exclusionMap[0, filtered_xy_locs[:, 1]-boundingBox[1], filtered_xy_locs[:, 0]-boundingBox[0]] = 1
```

To be honest, I just write this snippet in the comments and haven't checked if it works, but there should be a similar trick out there to avoid writing the inner for loop.

3- The smartest way maybe is to go with KDtree algorithm. I'm not an expert with this concept, but this might be the solution to both speed and memory efficiency.  For example, in the init method you form the KDtree to find the K nearest neighbors of each point and for each point when you are forming the exclusion map, you somehow extract the nearest neighbors that fall inside the same bounding box as well.

All and all, my suggestion is this: check for a high number of point annotations (like random 100000 points) and see how long does it take with your current exclusionMap creation function. If it was considerable, neglect it and go with the classic way that we are doing in orifinal NuClick i.e., creating a big click_map.